### PR TITLE
HPCC-12131 Xref Upgrade

### PR DESCRIPTION
**Pull request by miguelvazq**
Incorporate Xref into the new ecl watch paradigm.

Signed-off by: Miguel Vazquez miguel.vazquez@lexisnexis.com

### CODE REVIEW

**Comment by miguelvazq** [comment]
@GordonSmith Please review.

File: esp/src/eclwatch/WsDFUXref.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
+/*##############################################################################
+#    HPCC SYSTEMS software Copyright (C) 2012 HPCC Systems®.
```
Comment (GordonSmith) [inline esp/src/eclwatch/WsDFUXref.js]: Year?

Comment (GordonSmith) [inline esp/src/eclwatch/WsDFUXref.js]: Actually I believe we don't need this copyright message in each file any more as the single licence.txt should suffice?

File: esp/src/eclwatch/XrefDirectoriesWidget.js [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+                GridDetailsWidget, WsDFUXref, DelayLoadWidget, ESPUtil) {
+    return declare("XrefQueryWidget", [GridDetailsWidget], {
+        i18n: nlsHPCC,
+        filter: null,
+        filterLoaded: null,
+        gridTitle: nlsHPCC.title_DirectoriesFor,
+        idProperty: "Name",
+
+        init: function (params) {
+            var context = this;
```
Comment (GordonSmith) [inline esp/src/eclwatch/XrefDirectoriesWidget.js]: Context not used (I will be enabling this test with the linter soon!)

File: esp/src/eclwatch/XrefErrorsWarningsWidget.js [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+], function (declare, lang, i18n, nlsHPCC, arrayUtil, on, dom, domConstruct, domClass,
+                registry, ToggleButton, ToolbarSeparator, Button,
+                GridDetailsWidget, WsDFUXref, DelayLoadWidget, ESPUtil) {
+    return declare("XrefFoundFilesWidget", [GridDetailsWidget], {
+        i18n: nlsHPCC,
+        gridTitle: nlsHPCC.title_ErrorsWarnings,
+        idProperty: "Name",
+
+        init: function (params) {
+            var context = this;
```
Comment (GordonSmith) [inline esp/src/eclwatch/XrefErrorsWarningsWidget.js]: context not used

File: esp/src/eclwatch/XrefErrorsWarningsWidget.js [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        },
+
+        _onRefresh: function (event) {
+            this.refreshGrid();
+        },
+
+        createGrid: function (domID) {
+            this.openButton = registry.byId(this.id + "Open");
+
+            var context = this;
```
Comment (GordonSmith) [inline esp/src/eclwatch/XrefErrorsWarningsWidget.js]: context not used

File: esp/src/eclwatch/XrefErrorsWarningsWidget.js [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+                var results = [];
+                var newRows = [];
+                if (lang.exists("DFUXRefMessagesQueryResponse.DFUXRefMessagesQueryResult", response)) {
+                    results = response.DFUXRefMessagesQueryResponse.DFUXRefMessagesQueryResult;
+                }
+                arrayUtil.forEach(results.Warning, function (row, idx) {
+                   newRows.push({
+                        File: row.File,
+                        Text: row.Text,
+                        Status: context.i18n.Warning,
```
Comment (GordonSmith) [inline esp/src/eclwatch/XrefErrorsWarningsWidget.js]: trailing comma

File: esp/src/eclwatch/XrefFoundFilesWidget.js [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+], function (declare, lang, i18n, nlsHPCC, arrayUtil, on, dom, domConstruct, domClass,
+                registry, ToggleButton, ToolbarSeparator, Button,
+                GridDetailsWidget, WsDFUXref, DelayLoadWidget, ESPUtil) {
+    return declare("XrefFoundFilesWidget", [GridDetailsWidget], {
+        i18n: nlsHPCC,
+        gridTitle: nlsHPCC.title_FoundFilesFor,
+        idProperty: "Name",
+
+        init: function (params) {
+            var context = this;
```
Comment (GordonSmith) [inline esp/src/eclwatch/XrefFoundFilesWidget.js]: context not used

File: esp/src/eclwatch/XrefFoundFilesWidget.js [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+
+            this.gridTab.set("title", this.i18n.title_FoundFilesFor + this.params.Name);
+        },
+
+        _onRefresh: function (event) {
+            this.refreshGrid();
+        },
+
+        createGrid: function (domID) {
+            var context = this;
```
Comment (GordonSmith) [inline esp/src/eclwatch/XrefFoundFilesWidget.js]: context not used

File: esp/src/eclwatch/XrefLostFilesWidget.js [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+                registry, ToggleButton, ToolbarSeparator, Button,
+                selector,
+                GridDetailsWidget, WsDFUXref, ESPWorkunit, DelayLoadWidget, ESPUtil, FilterDropDownWidget, XrefDetailsWidget) {
+    return declare("XrefFoundFilesWidget", [GridDetailsWidget], {
+        i18n: nlsHPCC,
+        gridTitle: nlsHPCC.title_LostFilesFor,
+        idProperty: "Name",
+
+        init: function (params) {
+            var context = this;
```
Comment (GordonSmith) [inline esp/src/eclwatch/XrefLostFilesWidget.js]: context not used

File: esp/src/eclwatch/XrefLostFilesWidget.js [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+            this.gridTab.set("title", this.i18n.title_LostFilesFor + this.params.Name);
+        },
+
+        _onRefresh: function (event) {
+            this.refreshGrid();
+        },
+
+        createGrid: function (domID) {
+
+            var context = this;
```
Comment (GordonSmith) [inline esp/src/eclwatch/XrefLostFilesWidget.js]: context not used.

File: esp/src/eclwatch/XrefOrphanFilesWidget.js [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+                registry, ToggleButton, ToolbarSeparator, Button,
+                selector,
+                GridDetailsWidget, WsDFUXref, DelayLoadWidget, ESPUtil) {
+    return declare("XrefFoundFilesWidget", [GridDetailsWidget], {
+        i18n: nlsHPCC,
+        gridTitle: nlsHPCC.title_OrphanFilesFor,
+        idProperty: "Name",
+
+        init: function (params) {
+            var context = this;
```
Comment (GordonSmith) [inline esp/src/eclwatch/XrefOrphanFilesWidget.js]: context not used.

File: esp/src/eclwatch/XrefOrphanFilesWidget.js [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+            this.refreshGrid();
+        },
+
+        createGrid: function (domID) {
+            this.openButton = registry.byId(this.id + "Open");
+            this.delete = new Button({
+                id: this.id + "Delete",
+                disabled: false,
+                onClick: function (val) {
+                    if (confirm(context.i18n.RunningServerStrain)) {
```
Comment (GordonSmith) [inline esp/src/eclwatch/XrefOrphanFilesWidget.js]: context not declared.

File: esp/src/eclwatch/XrefOrphanFilesWidget.js [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+                        var selections = context.grid.getSelected();
+                        WsDFUXref.DFUXRefBuild({
+                            request: {
+                                Cluster: selections[0].Name
+                            }
+                        }).then(function (response) {
+                            if (response) {
+                                context.refreshGrid();
+                            }
+                        })
```
Comment (GordonSmith) [inline esp/src/eclwatch/XrefOrphanFilesWidget.js]: missing ;

File: esp/src/eclwatch/XrefOrphanFilesWidget.js [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+
+                    }
+
+                },
+                label: this.i18n.Delete
+            }).placeAt(this.openButton.domNode, "after");
+
+            var context = this;
```
Comment (GordonSmith) [inline esp/src/eclwatch/XrefOrphanFilesWidget.js]: Ahh here it is  - should move to top of procedure.

File: esp/src/eclwatch/XrefQueryWidget.js [hunk 14]
```diff
@@ -0,0 +1,217 @@
...
+                registry, ToggleButton, ToolbarSeparator, Button,
+                selector,
+                GridDetailsWidget, WsDFUXref, DelayLoadWidget, ESPUtil) {
+    return declare("XrefQueryWidget", [GridDetailsWidget], {
+        i18n: nlsHPCC,
+        gridTitle: nlsHPCC.XRef,
+        idProperty: "Name",
+
+        init: function (params) {
+            var context = this;
```
Comment (GordonSmith) [inline esp/src/eclwatch/XrefQueryWidget.js]: context not used

File: esp/src/eclwatch/nls/hpcc.js [hunk 15]
```diff
@@ -48,6 +48,8 @@ define({root:
     BlobPrefix: "BLOB Prefix",
     Bottom: "Bottom",
     Cancel: "Cancel",
+    CancelAll: "Cancel All",
+    CancelAllMessage: "This will attempt to stop any running jobs and clear the job queue. Do you wish to continue?",
```
Comment (GordonSmith) [inline esp/src/eclwatch/nls/hpcc.js]: How about "Abort running jobs and clear queue.  Do you wish to continue?"

File: esp/src/eclwatch/nls/hpcc.js [hunk 16]
```diff
@@ -95,6 +97,8 @@ define({root:
     Defaults: "Defaults",
     Delete: "Delete",
     Deleted: "Deleted",
+    DeleteEmptyDirectories: "Delete empty directories?",
+    DeleteDirectories: "You are about to delete empty directories. Do you wish to continue?",
```
Comment (GordonSmith) [inline esp/src/eclwatch/nls/hpcc.js]: "Remove empty directories.  Do you with to continue?"

File: esp/src/eclwatch/nls/hpcc.js [hunk 17]
```diff
@@ -196,10 +200,13 @@ define({root:
     Format: "Format",
     Forums: "Forums",
     Forward: "Forward",
+    FoundFile2: "Found File:",
+    FoundFileMessage: "A found file has all of its parts on disk that are not referenced in the Dali server. All the file parts are accounted for so they can be added back to the Dali server. They can also be deleted from the cluster, if required.",
```
Comment (GordonSmith) [inline esp/src/eclwatch/nls/hpcc.js]: Does this text come from the legacy version?  @JamesDeFabia do you want some input on these messages?

File: esp/src/eclwatch/nls/hpcc.js [hunk 18]
```diff
@@ -248,10 +255,14 @@ define({root:
     Largest: "Largest",
     LargestFile: "Largest File",
     LargestSize: "Largest Size",
+    LastMessage: "Last Message",
+    LastMessage2: "Last Message:",
```
Comment (GordonSmith) [inline esp/src/eclwatch/nls/hpcc.js]: Could we move the ":" into the JS?

File: esp/src/eclwatch/nls/hpcc.js [hunk 19]
```diff
@@ -248,10 +255,14 @@ define({root:
...
     LastName: "Last Name",
     LastNDays: "Last N Days",
     LastNHours: "Last N Hours",
     LastNRows: "Last N Rows",
+    LastRun: "Last Run",
+    LastRun2: "Last Run:",
```
Comment (GordonSmith) [inline esp/src/eclwatch/nls/hpcc.js]: Could we move the ":" into the JS?

File: esp/src/eclwatch/nls/hpcc.js [hunk 20]
```diff
@@ -277,6 +288,8 @@ define({root:
     Log: "Log",
     Logs: "Logs",
     log_analysis_1: "log_analysis_1*",
+    LostFile2: "Lost File:",
+    LostFileMessage: "A logical file that is missing at least one file part on both the primary and replicated locations in storage. The logical file is still referenced in the Dali server. Deleting the file removes the reference from the Dali server and any remaining parts on disk.",
```
Comment (GordonSmith) [inline esp/src/eclwatch/nls/hpcc.js]: @JamesDeFabia Please review

File: esp/src/eclwatch/nls/hpcc.js [hunk 21]
```diff
@@ -344,6 +361,8 @@ define({root:
     Operations: "Operations",
     Options: "Options",
     OriginalFile: "Original File",
+    OrphanFile2: "Orphan File:",
+    OrphanMessage: "An orphan file has partial file parts on disk. However, a full set of parts is not available to construct a complete logical file. These partial file parts, therefore, do not have a reference in the Dali server.",
```
Comment (GordonSmith) [inline esp/src/eclwatch/nls/hpcc.js]: @JamesDeFabia Please review

File: esp/src/eclwatch/nls/hpcc.js [hunk 22]
```diff
@@ -465,6 +489,7 @@ define({root:
     RowPath: "Row Path",
     RowTag: "Row Tag",
     RoxieCluster: "Roxie Cluster",
+    RunningServerStrain: "Running this process may take a long time and will put a heavy strain on the servers. Do you wish to continue?",
```
Comment (GordonSmith) [inline esp/src/eclwatch/nls/hpcc.js]: @JamesDeFabia Please Review

File: esp/src/eclwatch/templates/XrefDetailsWidget.html [hunk 23]
```diff
@@ -0,0 +1,52 @@
...
+                            </li>
+                            <li>
+                                <label class="Prompt" for="${id}LostFile">${i18n.LostFile2}</label>
+                                <div id="${id}LostFile">${i18n.LostFileMessage}</div>
+                            </li>
+                        </ul>
+                    </form>
+                </div>
+            </div>
+            <div id="${id}_FoundFiles" title="Found Files" data-dojo-props="delayWidget: 'XrefFoundFilesWidget'" data-dojo-type="DelayLoadWidget">
```
Comment (GordonSmith) [inline esp/src/eclwatch/templates/XrefDetailsWidget.html]: translations

File: esp/src/eclwatch/templates/XrefDetailsWidget.html [hunk 24]
```diff
@@ -0,0 +1,52 @@
...
+            </div>
+            <div id="${id}_OrphanFiles" title="Orphan Files" data-dojo-props="delayWidget: 'XrefOrphanFilesWidget'" data-dojo-type="DelayLoadWidget">
```
Comment (GordonSmith) [inline esp/src/eclwatch/templates/XrefDetailsWidget.html]: translations

File: esp/src/eclwatch/templates/XrefDetailsWidget.html [hunk 25]
```diff
@@ -0,0 +1,52 @@
...
+            </div>
+            <div id="${id}_LostFiles" title="Lost Files" data-dojo-props="delayWidget: 'XrefLostFilesWidget'" data-dojo-type="DelayLoadWidget">
```
Comment (GordonSmith) [inline esp/src/eclwatch/templates/XrefDetailsWidget.html]: translations

File: esp/src/eclwatch/WsDFUXref.js (hunk 1 above)
Comment (richardkchapman) [inline esp/src/eclwatch/WsDFUXref.js]: @fvillanustre Can you comment on what our policy on copyright messages in files needs to be?

File: esp/src/eclwatch/templates/XrefDetailsWidget.html [hunk 26]
```diff
@@ -0,0 +1,52 @@
...
+            </div>
+            <div id="${id}_Directories" title="Directories" data-dojo-props="delayWidget: 'XrefDirectoriesWidget'" data-dojo-type="DelayLoadWidget">
+            </div>
+            <div id="${id}_Errors" title="Errors/Warnings" data-dojo-props="delayWidget: 'XrefErrorsWarningsWidget'" data-dojo-type="DelayLoadWidget">
```
Comment (GordonSmith) [inline esp/src/eclwatch/templates/XrefDetailsWidget.html]: translations

File: esp/src/eclwatch/templates/XrefDetailsWidget.html [hunk 27]
```diff
@@ -0,0 +1,52 @@
...
+                    </form>
+                </div>
+            </div>
+            <div id="${id}_FoundFiles" title="Found Files" data-dojo-props="delayWidget: 'XrefFoundFilesWidget'" data-dojo-type="DelayLoadWidget">
+            </div>
+            <div id="${id}_OrphanFiles" title="Orphan Files" data-dojo-props="delayWidget: 'XrefOrphanFilesWidget'" data-dojo-type="DelayLoadWidget">
+            </div>
+            <div id="${id}_LostFiles" title="Lost Files" data-dojo-props="delayWidget: 'XrefLostFilesWidget'" data-dojo-type="DelayLoadWidget">
+            </div>
+            <div id="${id}_Directories" title="Directories" data-dojo-props="delayWidget: 'XrefDirectoriesWidget'" data-dojo-type="DelayLoadWidget">
```
Comment (GordonSmith) [inline esp/src/eclwatch/templates/XrefDetailsWidget.html]: translations

File: esp/src/eclwatch/templates/XrefErrorsWarningsWidget.js [hunk 28]
```diff
@@ -0,0 +1,270 @@
...
+                if (item.StatusReports && item.StatusReports.StatusReport && item.StatusReports.StatusReport.length) {
+                    return true;
+                }
+                return false;
+            };
+
+            this.store.getChildren = function (parent, options) {
+               var retVal =  this.query({__hpcc_parentName: parent.__hpcc_id}, options);
+               return retVal;
+            };*/
```
Comment (GordonSmith) [inline esp/src/eclwatch/templates/XrefErrorsWarningsWidget.js]: remove commented code.

File: esp/src/eclwatch/templates/XrefErrorsWarningsWidget.js [hunk 29]
```diff
@@ -0,0 +1,270 @@
...
+
+            this.openButton = registry.byId(this.id + "Open");
+            this.filter = new FilterDropDownWidget();
+            this.filter.placeAt(this.openButton.domNode, "after");
+
+            this.generate = new Button({
+                id: this.id + "Generate",
+                disabled: false,
+                onClick: function (val) {
+                    if (confirm("Running this process may take a long time and will put a heavy strain on the servers. Do you wish to continue?")) {
```
Comment (GordonSmith) [inline esp/src/eclwatch/templates/XrefErrorsWarningsWidget.js]: translation

File: esp/src/eclwatch/templates/XrefErrorsWarningsWidget.js [hunk 30]
```diff
@@ -0,0 +1,270 @@
...
+            }).placeAt(this.lostFiles.domNode, "after");
+            
+            this.errorWarnings = new Button({
+                id: this.id + "ErrorWarnings",
+                disabled: false,
+                onClick: function (val) {
+                    context.dialog.show();
+                },
+                label: "Errors/Warnings"
+             }).placeAt(this.directories.domNode, "after");*/
```
Comment (GordonSmith) [inline esp/src/eclwatch/templates/XrefErrorsWarningsWidget.js]: remove commented code

File: esp/src/eclwatch/templates/XrefErrorsWarningsWidget.js [hunk 31]
```diff
@@ -0,0 +1,270 @@
...
+            
+            new ToolbarSeparator().placeAt(this.openButton.domNode, "after");
+            new ToolbarSeparator().placeAt(this.cancel.domNode, "after");
+            /*new ToolbarSeparator().placeAt(this.errorWarnings.domNode, "after");
+            new ToolbarSeparator().placeAt(this.openButton.domNode, "after");
+            new ToolbarSeparator().placeAt(this.foundFiles.domNode, "before");*/
```
Comment (GordonSmith) [inline esp/src/eclwatch/templates/XrefErrorsWarningsWidget.js]: remove commented code.

File: esp/src/eclwatch/templates/XrefErrorsWarningsWidget.js [hunk 32]
```diff
@@ -0,0 +1,270 @@
...
+            }
+        },
+
+        getFilter: function () {
+            /*var retVal = this.filter.toObject();
+            
+            return retVal;
+            return {
+                Name: "SuperFiles"
+            }*/
```
Comment (GordonSmith) [inline esp/src/eclwatch/templates/XrefErrorsWarningsWidget.js]: remove commented code (is filter used in this widget?)

File: esp/src/eclwatch/templates/XrefErrorsWarningsWidget.js [hunk 33]
```diff
@@ -0,0 +1,270 @@
...
+                }
+                arrayUtil.forEach(results.XRefNode, function (row, idx) {
+                   newRows.push({
+                        Name: row.Name,
+                        Modified: row.Modified,
+                        Status: row.Status
+                    });
+                });
+                context.store.setData(newRows);
+                context.grid.set("query", /*this.getFilter()*/{});
```
Comment (GordonSmith) [inline esp/src/eclwatch/templates/XrefErrorsWarningsWidget.js]: remove commented code (and all refs to filters?)

File: esp/src/eclwatch/templates/XrefErrorsWarningsWidget.js [hunk 34]
```diff
@@ -0,0 +1,270 @@
...
+            });
+        },
+
+        ensurePane: function (id, params) {
+            id = this.createChildTabID(id);
+            var retVal = registry.byId(id);
+            if (!retVal) {
+                var context = this;
```
Comment (GordonSmith) [inline esp/src/eclwatch/templates/XrefErrorsWarningsWidget.js]: context not used.

File: esp/src/eclwatch/templates/XrefLostFilesWidget.js [hunk 35]
```diff
@@ -0,0 +1,270 @@
...
+                if (item.StatusReports && item.StatusReports.StatusReport && item.StatusReports.StatusReport.length) {
+                    return true;
+                }
+                return false;
+            };
+
+            this.store.getChildren = function (parent, options) {
+               var retVal =  this.query({__hpcc_parentName: parent.__hpcc_id}, options);
+               return retVal;
+            };*/
```
Comment (GordonSmith) [inline esp/src/eclwatch/templates/XrefLostFilesWidget.js]: remove commented code

File: esp/src/eclwatch/templates/XrefLostFilesWidget.js [hunk 36]
```diff
@@ -0,0 +1,270 @@
...
+
+            this.openButton = registry.byId(this.id + "Open");
+            this.filter = new FilterDropDownWidget();
+            this.filter.placeAt(this.openButton.domNode, "after");
+
+            this.generate = new Button({
+                id: this.id + "Generate",
+                disabled: false,
+                onClick: function (val) {
+                    if (confirm("Running this process may take a long time and will put a heavy strain on the servers. Do you wish to continue?")) {
```
Comment (GordonSmith) [inline esp/src/eclwatch/templates/XrefLostFilesWidget.js]: translation

File: esp/src/eclwatch/templates/XrefLostFilesWidget.js [hunk 37]
```diff
@@ -0,0 +1,270 @@
...
+            }).placeAt(this.lostFiles.domNode, "after");
+            
+            this.errorWarnings = new Button({
+                id: this.id + "ErrorWarnings",
+                disabled: false,
+                onClick: function (val) {
+                    context.dialog.show();
+                },
+                label: "Errors/Warnings"
+             }).placeAt(this.directories.domNode, "after");*/
```
Comment (GordonSmith) [inline esp/src/eclwatch/templates/XrefLostFilesWidget.js]: remove commented code.

File: esp/src/eclwatch/templates/XrefLostFilesWidget.js [hunk 38]
```diff
@@ -0,0 +1,270 @@
...
+            
+            new ToolbarSeparator().placeAt(this.openButton.domNode, "after");
+            new ToolbarSeparator().placeAt(this.cancel.domNode, "after");
+            /*new ToolbarSeparator().placeAt(this.errorWarnings.domNode, "after");
+            new ToolbarSeparator().placeAt(this.openButton.domNode, "after");
+            new ToolbarSeparator().placeAt(this.foundFiles.domNode, "before");*/
```
Comment (GordonSmith) [inline esp/src/eclwatch/templates/XrefLostFilesWidget.js]: remove commented code.

File: esp/src/eclwatch/templates/XrefLostFilesWidget.js [hunk 39]
```diff
@@ -0,0 +1,270 @@
...
+            }
+        },
+
+        getFilter: function () {
+            /*var retVal = this.filter.toObject();
+            
+            return retVal;
+            return {
+                Name: "SuperFiles"
+            }*/
```
Comment (GordonSmith) [inline esp/src/eclwatch/templates/XrefLostFilesWidget.js]: remove commented code.

File: esp/src/eclwatch/templates/XrefLostFilesWidget.js [hunk 40]
```diff
@@ -0,0 +1,270 @@
...
+                }
+                arrayUtil.forEach(results.XRefNode, function (row, idx) {
+                   newRows.push({
+                        Name: row.Name,
+                        Modified: row.Modified,
+                        Status: row.Status
+                    });
+                });
+                context.store.setData(newRows);
+                context.grid.set("query", /*this.getFilter()*/{});
```
Comment (GordonSmith) [inline esp/src/eclwatch/templates/XrefLostFilesWidget.js]: remove commentedd code and all refs to filters?

**Review by GordonSmith — CHANGES_REQUESTED** [review]

File: esp/src/eclwatch/XrefDetailsWidget.js [hunk 41]
```diff
@@ -0,0 +1,177 @@
...
+                dom.byId(this.id + "Cluster").textContent = params.Name;
+                dom.byId(this.id + "LastRun").textContent = params.Modified;
+                dom.byId(this.id + "LastMessage").textContent = params.Status;
+            }
+            this.refreshActionState();
+        },
+
+        initTab: function () {
+            var currSel = this.getSelectedChild();
+            if (currSel.id == this.widget._FoundFiles.id && !this.widget._FoundFiles.__hpcc_initalized) {
```
Comment (miguelvazq) [inline esp/src/eclwatch/XrefDetailsWidget.js]: I started seeing those in your code review and the `for (key` will now == `for (var key`

File: esp/src/eclwatch/nls/hpcc.js (hunk 16 above)
Comment (JamesDeFabia) [inline esp/src/eclwatch/nls/hpcc.js]: "Remove empty directories. Do you wish to continue?"

File: esp/src/eclwatch/nls/hpcc.js (hunk 20 above)
Comment (JamesDeFabia) [inline esp/src/eclwatch/nls/hpcc.js]: this text is fine

File: esp/src/eclwatch/nls/hpcc.js (hunk 21 above)
Comment (JamesDeFabia) [inline esp/src/eclwatch/nls/hpcc.js]: "An orphan file has partial file parts on disk. However, a full set of parts is not available to construct a complete logical file. This no reference to these file parts in the Dali server."

File: esp/src/eclwatch/nls/hpcc.js (hunk 22 above)
Comment (JamesDeFabia) [inline esp/src/eclwatch/nls/hpcc.js]: This text is ok

**Comment by richardkchapman** [comment]
@miguelvazq Back to you

File: esp/src/eclwatch/nls/hpcc.js (hunk 18 above)
Comment (miguelvazq) [inline esp/src/eclwatch/nls/hpcc.js]: How about in the html I can do something like `<label class="Prompt" for="${id}LastMessage">${i18n.LastMessage2}:</label>`

**Comment by miguelvazq** [comment]
@GordonSmith Updated

File: esp/src/eclwatch/nls/hpcc.js [hunk 42]
```diff
@@ -196,10 +201,14 @@ define({root:
     Format: "Format",
     Forums: "Forums",
     Forward: "Forward",
+    FoundFile: "Found Files",
+    FoundFile2: "Found File:",
```
Comment (GordonSmith) [inline esp/src/eclwatch/nls/hpcc.js]: This can be removed

File: esp/src/eclwatch/nls/hpcc.js [hunk 43]
```diff
@@ -277,6 +290,9 @@ define({root:
     Log: "Log",
     Logs: "Logs",
     log_analysis_1: "log_analysis_1*",
+    LostFile: "Lost Files:",
+    LostFile2: "Lost File:",
```
Comment (GordonSmith) [inline esp/src/eclwatch/nls/hpcc.js]: Are the ":" going to be moved into the HTML?

File: esp/src/eclwatch/nls/hpcc.js [hunk 44]
```diff
@@ -344,6 +364,9 @@ define({root:
     Operations: "Operations",
     Options: "Options",
     OriginalFile: "Original File",
+    OrphanFile: "Orphan Files:",
+    OrphanFile2: "Orphan File:",
```
Comment (GordonSmith) [inline esp/src/eclwatch/nls/hpcc.js]: Are the ":" going to be moved into the HTML?

File: esp/src/eclwatch/nls/hpcc.js [hunk 45]
```diff
@@ -556,12 +587,15 @@ define({root:
     title_AvailableGroupPermissions: "Available Group Permissions",
     title_ClusterInfo: "Groups",
     title_CodeGeneratorPermissions: "Code Generator Permissions",
+    title_DirectoriesFor: "Directories for:",
```
Comment (GordonSmith) [inline esp/src/eclwatch/nls/hpcc.js]: Are the ":" going to be moved into the HTML?

File: esp/src/eclwatch/nls/hpcc.js [hunk 46]
```diff
@@ -556,12 +587,15 @@ define({root:
...
     title_DFUQuery: "Logical Files",
     title_DFUWUDetails: "DFU Workunit",
     title_DiskUsage: "Disk Usage",
     title_ECLPlayground: "ECL Playground",
+    title_ErrorsWarnings: "Errors/Warnings for:",
```
Comment (GordonSmith) [inline esp/src/eclwatch/nls/hpcc.js]: Are the ":" going to be moved into the HTML?

File: esp/src/eclwatch/nls/hpcc.js [hunk 47]
```diff
@@ -556,12 +587,15 @@ define({root:
...
     title_EventScheduleWorkunit: "Event Scheduler",
     title_FileScopeDefaultPermissions: "Default permissions of files",
+    title_FoundFilesFor: "Found files for:",
```
Comment (GordonSmith) [inline esp/src/eclwatch/nls/hpcc.js]: Are the ":" going to be moved into the HTML?

File: esp/src/eclwatch/nls/hpcc.js [hunk 48]
```diff
@@ -579,7 +613,9 @@ define({root:
     title_LZBrowse: "Landing Zones",
     title_MemberOf: "Member Of",
     title_Members: "Members",
+    title_LostFilesFor: "Lost files for:",
```
Comment (GordonSmith) [inline esp/src/eclwatch/nls/hpcc.js]: Are the ":" going to be moved into the HTML?

File: esp/src/eclwatch/nls/hpcc.js [hunk 49]
```diff
@@ -579,7 +613,9 @@ define({root:
...
     title_LibrariesUsed: "Libraries Used",
+    title_OrphanFilesFor: "Orphan files for:",
```
Comment (GordonSmith) [inline esp/src/eclwatch/nls/hpcc.js]: Are the ":" going to be moved into the HTML?

File: esp/src/eclwatch/nls/hpcc.js [hunk 50]
```diff
@@ -248,10 +257,14 @@ define({root:
     Largest: "Largest",
     LargestFile: "Largest File",
     LargestSize: "Largest Size",
+    LastMessage: "Last Message",
+    LastMessage2: "Last Message",
```
Comment (GordonSmith) [inline esp/src/eclwatch/nls/hpcc.js]: LastMessage2 Not needed (same as LastMessage)

File: esp/src/eclwatch/nls/hpcc.js [hunk 51]
```diff
@@ -248,10 +257,14 @@ define({root:
...
     LastName: "Last Name",
     LastNDays: "Last N Days",
     LastNHours: "Last N Hours",
     LastNRows: "Last N Rows",
+    LastRun: "Last Run",
+    LastRun2: "Last Run",
```
Comment (GordonSmith) [inline esp/src/eclwatch/nls/hpcc.js]: LastRun2 duplicate of LastRun

**Review by GordonSmith — CHANGES_REQUESTED** [review]

**Comment by richardkchapman** [comment]
@miguelvazq Back to you

**Comment by miguelvazq** [comment]
@GordonSmith updated.

File: esp/src/eclwatch/XrefFoundFilesWidget.js [hunk 52]
```diff
@@ -0,0 +1,99 @@
...
+
+        createGrid: function (domID) {
+
+            var retVal = new declare([ESPUtil.Grid(true, true)])({
+                store: this.store,
+                columns: {
+                    Name: {label: this.i18n.Name, width:100, sortable: false},
+                    Modified: {label: this.i18n.Modified, width: 30, sortable: true},
+                    Parts: {label: this.i18n.Parts, width: 30, sortable: true},
+                    Size: {label: this.i18n.Size, width: 30, sortable: true},
```
Comment (GordonSmith) [inline esp/src/eclwatch/XrefFoundFilesWidget.js]: Trailing ","

File: esp/src/eclwatch/XrefOrphanFilesWidget.js [hunk 53]
```diff
@@ -0,0 +1,130 @@
...
+            this.refreshGrid();
+
+            this.gridTab.set("title", this.i18n.title_OrphanFilesFor + ":" + this.params.Name);
+        },
+
+        _onRefresh: function (event) {
+            this.refreshGrid();
+        },
+
+        createGrid: function (domID) {\
```
Comment (GordonSmith) [inline esp/src/eclwatch/XrefOrphanFilesWidget.js]: trailing "\"

File: esp/src/eclwatch/XrefQueryWidget.js [hunk 54]
```diff
@@ -0,0 +1,216 @@
...
+                        selectorType: 'checkbox',
+                        label: ""
+                    }),
+                    Name: {label: this.i18n.Name, width:100, sortable: false,
+                        formatter: function (Name, idx) {
+                           return "<a href='#' class='dgrid-row-url'>" + Name + "</a>";
+                        }
+                    },
+                    Modified: {label: this.i18n.LastRun, width: 30, sortable: false},
+                    Status: {label: this.i18n.LastMessage, width: 30, sortable: false},
```
Comment (GordonSmith) [inline esp/src/eclwatch/XrefQueryWidget.js]: trailing ","

File: esp/src/eclwatch/templates/XrefErrorsWarningsWidget.js (hunk 28 above)
Comment (GordonSmith) [inline esp/src/eclwatch/templates/XrefErrorsWarningsWidget.js]: Still needs removing

File: esp/src/eclwatch/templates/XrefErrorsWarningsWidget.js (hunk 29 above)
Comment (GordonSmith) [inline esp/src/eclwatch/templates/XrefErrorsWarningsWidget.js]: Still needs translation

File: esp/src/eclwatch/templates/XrefErrorsWarningsWidget.js (hunk 30 above)
Comment (GordonSmith) [inline esp/src/eclwatch/templates/XrefErrorsWarningsWidget.js]: Commented code still needs removing

File: esp/src/eclwatch/templates/XrefErrorsWarningsWidget.js (hunk 31 above)
Comment (GordonSmith) [inline esp/src/eclwatch/templates/XrefErrorsWarningsWidget.js]: Needs removing

File: esp/src/eclwatch/templates/XrefErrorsWarningsWidget.js (hunk 32 above)
Comment (GordonSmith) [inline esp/src/eclwatch/templates/XrefErrorsWarningsWidget.js]: commented code

File: esp/src/eclwatch/templates/XrefErrorsWarningsWidget.js (hunk 33 above)
Comment (GordonSmith) [inline esp/src/eclwatch/templates/XrefErrorsWarningsWidget.js]: Commented code.

File: esp/src/eclwatch/templates/XrefErrorsWarningsWidget.js (hunk 34 above)
Comment (GordonSmith) [inline esp/src/eclwatch/templates/XrefErrorsWarningsWidget.js]: Context can be removed?

**Review by GordonSmith — CHANGES_REQUESTED** [review]

**Comment by miguelvazq** [comment]
I removed an accidental file that was added into templates. The ones with the commented code comments. All else was updated.

**Review by GordonSmith — APPROVED** [review]

**Pull request merged**